src/java.base/share/classes/java/io/FileInputStream.java

```diff
@@ -236,14 +237,51 @@ public int read() throws IOException {
 
     private native int read0() throws IOException;
 
+    private native int readBytes0(byte[] b, int off, int len) throws IOException;
+
     /**
      * Reads a subarray as a sequence of bytes.
      * @param     b the data to be written
      * @param     off the start offset in the data
      * @param     len the number of bytes that are written
      * @throws    IOException If an I/O error has occurred.
      */
-    private native int readBytes(byte[] b, int off, int len) throws IOException;
+    private int readBytes(byte[] b, int off, int len) throws IOException {
+        Objects.checkFromIndexSize(off, len, b.length);
+        int nread = 0;
+        int pos = off;
+        int remaining = len;
+        long comp = Blocker.begin();
+        try {
+            do {
+                int size = Math.min(remaining, 1572864);
```

Comment:
1.5Mb seems high, I think we really need micro that do real file I/O to help tune this.